<?xml version="1.0" encoding="utf-8" ?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="Docutils 0.13: http://docutils.sourceforge.net/" />
<title>Imputing Sex in Hail</title>
<style type="text/css">

/*
:Author: David Goodger (goodger@python.org)
:Id: $Id: html4css1.css 7952 2016-07-26 18:15:59Z milde $
:Copyright: This stylesheet has been placed in the public domain.

Default cascading style sheet for the HTML output of Docutils.

See http://docutils.sf.net/docs/howto/html-stylesheets.html for how to
customize this style sheet.
*/

/* used to remove borders from tables and images */
.borderless, table.borderless td, table.borderless th {
  border: 0 }

table.borderless td, table.borderless th {
  /* Override padding for "table.docutils td" with "! important".
     The right padding separates the table cells. */
  padding: 0 0.5em 0 0 ! important }

.first {
  /* Override more specific margin styles with "! important". */
  margin-top: 0 ! important }

.last, .with-subtitle {
  margin-bottom: 0 ! important }

.hidden {
  display: none }

.subscript {
  vertical-align: sub;
  font-size: smaller }

.superscript {
  vertical-align: super;
  font-size: smaller }

a.toc-backref {
  text-decoration: none ;
  color: black }

blockquote.epigraph {
  margin: 2em 5em ; }

dl.docutils dd {
  margin-bottom: 0.5em }

object[type="image/svg+xml"], object[type="application/x-shockwave-flash"] {
  overflow: hidden;
}

/* Uncomment (and remove this text!) to get bold-faced definition list terms
dl.docutils dt {
  font-weight: bold }
*/

div.abstract {
  margin: 2em 5em }

div.abstract p.topic-title {
  font-weight: bold ;
  text-align: center }

div.admonition, div.attention, div.caution, div.danger, div.error,
div.hint, div.important, div.note, div.tip, div.warning {
  margin: 2em ;
  border: medium outset ;
  padding: 1em }

div.admonition p.admonition-title, div.hint p.admonition-title,
div.important p.admonition-title, div.note p.admonition-title,
div.tip p.admonition-title {
  font-weight: bold ;
  font-family: sans-serif }

div.attention p.admonition-title, div.caution p.admonition-title,
div.danger p.admonition-title, div.error p.admonition-title,
div.warning p.admonition-title, .code .error {
  color: red ;
  font-weight: bold ;
  font-family: sans-serif }

/* Uncomment (and remove this text!) to get reduced vertical space in
   compound paragraphs.
div.compound .compound-first, div.compound .compound-middle {
  margin-bottom: 0.5em }

div.compound .compound-last, div.compound .compound-middle {
  margin-top: 0.5em }
*/

div.dedication {
  margin: 2em 5em ;
  text-align: center ;
  font-style: italic }

div.dedication p.topic-title {
  font-weight: bold ;
  font-style: normal }

div.figure {
  margin-left: 2em ;
  margin-right: 2em }

div.footer, div.header {
  clear: both;
  font-size: smaller }

div.line-block {
  display: block ;
  margin-top: 1em ;
  margin-bottom: 1em }

div.line-block div.line-block {
  margin-top: 0 ;
  margin-bottom: 0 ;
  margin-left: 1.5em }

div.sidebar {
  margin: 0 0 0.5em 1em ;
  border: medium outset ;
  padding: 1em ;
  background-color: #ffffee ;
  width: 40% ;
  float: right ;
  clear: right }

div.sidebar p.rubric {
  font-family: sans-serif ;
  font-size: medium }

div.system-messages {
  margin: 5em }

div.system-messages h1 {
  color: red }

div.system-message {
  border: medium outset ;
  padding: 1em }

div.system-message p.system-message-title {
  color: red ;
  font-weight: bold }

div.topic {
  margin: 2em }

h1.section-subtitle, h2.section-subtitle, h3.section-subtitle,
h4.section-subtitle, h5.section-subtitle, h6.section-subtitle {
  margin-top: 0.4em }

h1.title {
  text-align: center }

h2.subtitle {
  text-align: center }

hr.docutils {
  width: 75% }

img.align-left, .figure.align-left, object.align-left, table.align-left {
  clear: left ;
  float: left ;
  margin-right: 1em }

img.align-right, .figure.align-right, object.align-right, table.align-right {
  clear: right ;
  float: right ;
  margin-left: 1em }

img.align-center, .figure.align-center, object.align-center {
  display: block;
  margin-left: auto;
  margin-right: auto;
}

table.align-center {
  margin-left: auto;
  margin-right: auto;
}

.align-left {
  text-align: left }

.align-center {
  clear: both ;
  text-align: center }

.align-right {
  text-align: right }

/* reset inner alignment in figures */
div.align-right {
  text-align: inherit }

/* div.align-center * { */
/*   text-align: left } */

.align-top    {
  vertical-align: top }

.align-middle {
  vertical-align: middle }

.align-bottom {
  vertical-align: bottom }

ol.simple, ul.simple {
  margin-bottom: 1em }

ol.arabic {
  list-style: decimal }

ol.loweralpha {
  list-style: lower-alpha }

ol.upperalpha {
  list-style: upper-alpha }

ol.lowerroman {
  list-style: lower-roman }

ol.upperroman {
  list-style: upper-roman }

p.attribution {
  text-align: right ;
  margin-left: 50% }

p.caption {
  font-style: italic }

p.credits {
  font-style: italic ;
  font-size: smaller }

p.label {
  white-space: nowrap }

p.rubric {
  font-weight: bold ;
  font-size: larger ;
  color: maroon ;
  text-align: center }

p.sidebar-title {
  font-family: sans-serif ;
  font-weight: bold ;
  font-size: larger }

p.sidebar-subtitle {
  font-family: sans-serif ;
  font-weight: bold }

p.topic-title {
  font-weight: bold }

pre.address {
  margin-bottom: 0 ;
  margin-top: 0 ;
  font: inherit }

pre.literal-block, pre.doctest-block, pre.math, pre.code {
  margin-left: 2em ;
  margin-right: 2em }

pre.code .ln { color: grey; } /* line numbers */
pre.code, code { background-color: #eeeeee }
pre.code .comment, code .comment { color: #5C6576 }
pre.code .keyword, code .keyword { color: #3B0D06; font-weight: bold }
pre.code .literal.string, code .literal.string { color: #0C5404 }
pre.code .name.builtin, code .name.builtin { color: #352B84 }
pre.code .deleted, code .deleted { background-color: #DEB0A1}
pre.code .inserted, code .inserted { background-color: #A3D289}

span.classifier {
  font-family: sans-serif ;
  font-style: oblique }

span.classifier-delimiter {
  font-family: sans-serif ;
  font-weight: bold }

span.interpreted {
  font-family: sans-serif }

span.option {
  white-space: nowrap }

span.pre {
  white-space: pre }

span.problematic {
  color: red }

span.section-subtitle {
  /* font-size relative to parent (h1..h6 element) */
  font-size: 80% }

table.citation {
  border-left: solid 1px gray;
  margin-left: 1px }

table.docinfo {
  margin: 2em 4em }

table.docutils {
  margin-top: 0.5em ;
  margin-bottom: 0.5em }

table.footnote {
  border-left: solid 1px black;
  margin-left: 1px }

table.docutils td, table.docutils th,
table.docinfo td, table.docinfo th {
  padding-left: 0.5em ;
  padding-right: 0.5em ;
  vertical-align: top }

table.docutils th.field-name, table.docinfo th.docinfo-name {
  font-weight: bold ;
  text-align: left ;
  white-space: nowrap ;
  padding-left: 0 }

/* "booktabs" style (no vertical lines) */
table.docutils.booktabs {
  border: 0px;
  border-top: 2px solid;
  border-bottom: 2px solid;
  border-collapse: collapse;
}
table.docutils.booktabs * {
  border: 0px;
}
table.docutils.booktabs th {
  border-bottom: thin solid;
  text-align: left;
}

h1 tt.docutils, h2 tt.docutils, h3 tt.docutils,
h4 tt.docutils, h5 tt.docutils, h6 tt.docutils {
  font-size: 100% }

ul.auto-toc {
  list-style-type: none }

</style>
<style type="text/css">

@import url(html4css1.css);

h1, h2, h3, h4, h5, h6, p.topic-title {
  font-family: arial }

h1.title {
	text-align: left
}

tt {
	color: red
}

table {
    border-collapse: collapse;
}

table, th, td {
    border: 1px solid black;
}

tr:nth-child(even) {background-color: #f2f2f2}
</style>
</head>
<body>
<div class="document" id="imputing-sex-in-hail">
<h1 class="title">Imputing Sex in Hail</h1>

<p>Hail contains an <tt class="docutils literal">imputesex</tt> module which will calculate the inbreeding coefficient on the X chromosome as done in <a class="reference external" href="http://pngu.mgh.harvard.edu/~purcell/plink/summary.shtml#sexcheck">PLINK v1.7</a>. All results are stored in sample annotations.</p>
<div class="section" id="command-line-options">
<h1>Command Line Options</h1>
<!-- This is using the option format. No way to put in long flags and short flags. Defaults must be written in description -->
<table class="docutils option-list" frame="void" rules="none">
<col class="option" />
<col class="description" />
<tbody valign="top">
<tr><td class="option-group">
<kbd><span class="option">-m <var>&lt;Double&gt;</var></span></kbd></td>
<td>Minimum variant minor allele frequency [Default=0.1]</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-i</span></kbd></td>
<td>Include variants in pseudo autosomal regions (HG19) [Default=false]</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-x <var>&lt;Double&gt;</var></span></kbd></td>
<td>If the inbreeding coefficient is less than <cite>&lt;Double&gt;</cite>, then sample is called as Female [Default=0.2]</td></tr>
<tr><td class="option-group">
<kbd><span class="option">-y <var>&lt;Double&gt;</var></span></kbd></td>
<td>If the inbreeding coefficient is greater than <cite>&lt;Double&gt;</cite>, then sample is called as Male [Default=0.8]</td></tr>
<tr><td class="option-group" colspan="2">
<kbd><span class="option">-p <var>&lt;expression&gt;</var></span></kbd></td>
</tr>
<tr><td>&nbsp;</td><td>Population minor allele frequency (variant annotation expression, e.g. <cite>va.exac.AF</cite>)</td></tr>
</tbody>
</table>
<!-- This is using the table format. Really annoying to format!!!! -->
<table border="1" class="docutils">
<colgroup>
<col width="31%" />
<col width="63%" />
<col width="6%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Flag</th>
<th class="head">Description</th>
<th class="head">Default</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt class="docutils literal"><span class="pre">-m</span> &lt;Double&gt; | <span class="pre">--maf-threshold</span> &lt;Double&gt;</tt></td>
<td>Minimum variant minor allele frequency</td>
<td>0.1</td>
</tr>
<tr><td><tt class="docutils literal"><span class="pre">-i</span> | <span class="pre">--include-par</span></tt></td>
<td>Include variants in pseudo autosomal regions (HG19)</td>
<td>False</td>
</tr>
<tr><td><tt class="docutils literal"><span class="pre">-x</span> &lt;Double&gt; | <span class="pre">--female-threshold</span> &lt;Double&gt;</tt></td>
<td>If the inbreeding coefficient is less than <cite>&lt;Double&gt;</cite>, then sample is called as Female</td>
<td>0.2</td>
</tr>
<tr><td><tt class="docutils literal"><span class="pre">-y</span> &lt;Double&gt; | <span class="pre">--male-threshold</span> &lt;Double&gt;</tt></td>
<td>If the inbreeding coefficient is greater than <cite>&lt;Double&gt;</cite>, then sample is called as Male</td>
<td>0.8</td>
</tr>
<tr><td><tt class="docutils literal"><span class="pre">-p</span> &lt;expression&gt; | <span class="pre">--pop-freq</span> &lt;expression&gt;</tt></td>
<td>Population minor allele frequency (variant annotation expression, e.g. <cite>va.exac.AF</cite>)</td>
<td>&nbsp;</td>
</tr>
</tbody>
</table>
<!-- Example table constructed using table generator (http://www.tablesgenerator.com/text_tables); won't compile! -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
<!-- | ``-m <Double> | - -maf-threshold <Double>       | Minimum variant minor allele frequency                                                    | 0.1   | -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
<!-- | ``-i | - -include-par``                         | Include variants in pseudo autosomal regions (HG19)                                       | False | -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
<!-- | ``-x <Double> | - -female-threshold <Double> `` | If the inbreeding coefficient is less than ``<Double>``, then sample is called as Female  | 0.2   | -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
<!-- | ``-y <Double> | - -male-threshold <Double>``    | If the inbreeding coefficient is greater than ``<Double>``, then sample is called as Male | 0.8   | -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
<!-- | ``-p <expression> | - -pop-freq <expression> `` | Population minor allele frequency (variant annotation expression, e.g. ``va.exac.AF``)    |       | -->
<!-- +- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -+- - - - - - -+ -->
</div>
<div class="section" id="example-imputesex-command">
<h1>Example <cite>imputesex</cite> Command</h1>
<pre class="code literal-block">
hail read -i /path/to/file.vds imputesex -m 0.01 exportsamples -o /path/to/output.tsv -c &quot;ID=s.id, Fstat=sa.imputesex.Fstat, ImputedSex=sa.imputesex.isFemale&quot;
</pre>
<p>To get the exact same answer as PLINK, make sure you split multi-allelic variants and use the flag <tt class="docutils literal"><span class="pre">--include-par</span></tt>:</p>
<pre class="code literal-block">
hail read -i /path/to/file.vds splitmulti imputesex --include-par
</pre>
</div>
<div class="section" id="implementation-details">
<h1>Implementation Details</h1>
<ol class="arabic simple">
<li>X chromosome variants are selected from the VDS: <tt class="docutils literal">v.contig == &quot;X&quot; || v.contig == &quot;23&quot;</tt></li>
<li>Variants with a minor allele frequency less than the threshold given by <tt class="docutils literal"><span class="pre">--mafthreshold</span></tt> are removed</li>
<li>Variants in the pseudoautosomal region <tt class="docutils literal"><span class="pre">(X:60001-2699520)</span> || <span class="pre">(X:154931044-155260560)</span></tt> are included if the <tt class="docutils literal"><span class="pre">--include-par</span></tt> flag is set</li>
<li>The minor allele frequency (maf) per variant is calculated</li>
<li>For each variant and sample with a non-missing genotype call, <tt class="docutils literal">E</tt>, the expected number of homozygotes (from population MAF), is computed as <tt class="docutils literal">1.0 - <span class="pre">(2.0*maf*(1.0-maf))</span></tt></li>
<li>For each variant and sample with a non-missing genotype call, <tt class="docutils literal">O</tt>, the observed number of homozygotes, is computed as <tt class="docutils literal">0 = heterozygote; 1 = homozygote</tt></li>
<li>For each variant and sample with a non-missing genotype call, <tt class="docutils literal">N</tt> is incremented by 1</li>
<li>For each sample, <tt class="docutils literal">E</tt>, <tt class="docutils literal">O</tt>, and <tt class="docutils literal">N</tt> are combined across variants</li>
<li><tt class="docutils literal">F</tt> is calculated by <tt class="docutils literal">(O - E) / (N - E)</tt></li>
<li>A sex is assigned to each sample with the following criteria: <tt class="docutils literal">F &lt; 0.2 =&gt; Female; F &gt; 0.8 =&gt; Male</tt>. Use <tt class="docutils literal"><span class="pre">--female-threshold</span></tt> and <tt class="docutils literal"><span class="pre">--male-threshold</span></tt> to change this behavior.</li>
</ol>
</div>
<div class="section" id="available-sample-annotations">
<h1>Available sample annotations</h1>
<p>The below annotations can be accessed with <tt class="docutils literal"><span class="pre">sa.imputesex.&lt;identifier&gt;</span></tt></p>
<table border="1" class="docutils">
<colgroup>
<col width="16%" />
<col width="11%" />
<col width="74%" />
</colgroup>
<thead valign="bottom">
<tr><th class="head">Identifier</th>
<th class="head">Type</th>
<th class="head">Description</th>
</tr>
</thead>
<tbody valign="top">
<tr><td><tt class="docutils literal">isFemale</tt></td>
<td><tt class="docutils literal">Boolean</tt></td>
<td>True if the imputed sex is female, false if male, missing if undetermined</td>
</tr>
<tr><td><tt class="docutils literal">Fstat</tt></td>
<td><tt class="docutils literal">Double</tt></td>
<td>Inbreeding coefficient</td>
</tr>
<tr><td><tt class="docutils literal">nTotal</tt></td>
<td><tt class="docutils literal">Int</tt></td>
<td>Total number of variants considered</td>
</tr>
<tr><td><tt class="docutils literal">nCalled</tt></td>
<td><tt class="docutils literal">Int</tt></td>
<td>Number of variants with a genotype call</td>
</tr>
<tr><td><tt class="docutils literal">expectedHoms</tt></td>
<td><tt class="docutils literal">Double</tt></td>
<td>Expected number of homozygotes</td>
</tr>
<tr><td><tt class="docutils literal">observedHoms</tt></td>
<td><tt class="docutils literal">Double</tt></td>
<td>Observed number of homozygotes</td>
</tr>
</tbody>
</table>
</div>
</div>
</body>
</html>
