<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>
</head>
<body>
<h1 id="pca-in-hail">PCA in Hail</h1>
<p>Hail supports principal component analysis (PCA) of genotype data, a now-standard procedure (<a href="http://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.0020190">Patterson, Price and Reich, 2006</a>).</p>
<p>As input PCA expects a <code>.vds</code> file with bialellic autosomal variants. The analysis is with respect to a standardized genotype matrix; see below for details. The default output is a <code>.tsv</code> file recording the first 10 principal component (PC) scores of each sample. Format: one line per sample plus a header, with columns <code>SAMPLE</code>, <code>PC1</code>, <code>PC2</code>, etc.</p>
<p>Example usage:</p>
<pre><code>$ hail read -i /path/to/file.vds pca -o /path/to/file.tsv</code></pre>
<p>Command line options: - <code>-k &lt;k&gt;</code>, <code>--components &lt;k&gt;</code> -- Report the first <span class="math inline">\(k\)</span> PCs; by default <span class="math inline">\(k = 10\)</span>. - <code>-l &lt;/path/to/file.tsv&gt;</code>, <code>--loadings</code> -- Write out the variant loadings for the first <span class="math inline">\(k\)</span> PCs. Format: one line per variant plus a header, with columns <code>CHROM POS REF ALT PC1 PC2</code> etc. - <code>-e &lt;/path/to/file.tsv&gt;</code>, <code>--eigenvalues</code> -- Write out the first <span class="math inline">\(k\)</span> eigenvalues of the sample correlation or genetic relationship matrix. Format: single column, no header.</p>
<h2 id="details">Details</h2>
<p>PCA is based on the singular value decomposition (SVD) of a standardized genotype matrix <span class="math inline">\(M\)</span>, computed as follows.</p>
<p>An <span class="math inline">\(n\times m\)</span> matrix <span class="math inline">\(C\)</span> records raw genotypes, with rows indexed by <span class="math inline">\(n\)</span> samples and columns indexed by <span class="math inline">\(m\)</span> bialellic autosomal variants; <span class="math inline">\(C_{ij}\)</span> is the number of alternate alleles of variant <span class="math inline">\(j\)</span> carried by sample <span class="math inline">\(i\)</span>, which can be 0, 1, 2, or missing. For each variant <span class="math inline">\(j\)</span>, the sample alternate allele frequency <span class="math inline">\(p_j\)</span> is computed as half the mean of the non-missing entries of column <span class="math inline">\(j\)</span>. Entries of <span class="math inline">\(M\)</span> are then mean-centered and variance-normalized as <span class="math display">\[
M_{ij} = \frac{C_{ij}-2p_j}{\sqrt{2p_j(1-p_j)m}},
\]</span> with <span class="math inline">\(M_{ij} = 0\)</span> for <span class="math inline">\(C_{ij}\)</span> missing (i.e. mean genotype imputation). This scaling normalizes genotype variances to a common value <span class="math inline">\(1/m\)</span> for variants in Hardy-Weinberg equilibrium and is further motivated in the paper cited above. (The resulting amplification of signal from the low end of the allele frequency spectrum will also introduce noise for rare variants; common practice is to filter out variants with minor allele frequency below some cutoff.) The factor <span class="math inline">\(1/m\)</span> gives each sample row approximately unit total variance (assuming linkage equilibrium) and yields the sample correlation or genetic relationship matrix (GRM) as simply <span class="math inline">\(MM^T\)</span>.</p>
<p>PCA then computes the SVD <span class="math display">\[
M = USV^T
\]</span> where columns of <span class="math inline">\(U\)</span> are left singular vectors (orthonormal in <span class="math inline">\(\mathbb{R}^n\)</span>), columns of <span class="math inline">\(V\)</span> are right singular vectors (orthonormal in <span class="math inline">\(\mathbb{R}^m\)</span>), and <span class="math inline">\(S=\mathrm{diag}(s_1, s_2, \ldots)\)</span> with ordered singular values <span class="math inline">\(s_1 \ge s_2 \ge \cdots \ge 0\)</span>. Typically one computes only the first <span class="math inline">\(k\)</span> singular vectors and values, yielding the best rank <span class="math inline">\(k\)</span> approximation <span class="math inline">\(U_k S_k V_k^T\)</span> of <span class="math inline">\(M\)</span>; the truncations <span class="math inline">\(U_k\)</span>, <span class="math inline">\(S_k\)</span> and <span class="math inline">\(V_k\)</span> are <span class="math inline">\(n\times k\)</span>, <span class="math inline">\(k\times k\)</span> and <span class="math inline">\(m\times k\)</span> respectively.</p>
<p>From the perspective of the samples or rows of <span class="math inline">\(M\)</span> as data, <span class="math inline">\(V_k\)</span> contains the variant loadings for the first <span class="math inline">\(k\)</span> PCs while <span class="math inline">\(MV_k = U_k S_k\)</span> contains the first <span class="math inline">\(k\)</span> PC scores of each sample. The loadings represent a new basis of features while the scores represent the projected data on those features. The eigenvalues of the GRM <span class="math inline">\(MM^T\)</span> are the squares of the singular values <span class="math inline">\(s_1^2, s_2^2, \ldots\)</span>, which represent the variances carried by the respective PCs.</p>
<p><strong>Note:</strong> In PLINK/GCTA the GRM is taken as the starting point and it is computed slightly differently with regard to missing data. Here the <span class="math inline">\(ij\)</span> entry of <span class="math inline">\(MM^T\)</span> is simply the dot product of rows <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> of <span class="math inline">\(M\)</span>; in terms of <span class="math inline">\(C\)</span> it is <span class="math display">\[
\frac{1}{m}\sum{l\in\mathcal{C}_i\cap\mathcal{C}_j}\frac{(C_{il}-2p_l)(C_{jl} - 2p_l)}{2p_l(1-p_l)}
\]</span> where <span class="math inline">\(\mathcal{C}_i = \{l\mid C_{il}\text{ is non-missing}\}\)</span>. In PLINK/GCTA the denominator <span class="math inline">\(m\)</span> is replaced with the number of terms in the sum <span class="math inline">\(\lvert\mathcal{C}_i\cap\mathcal{C}_j\rvert\)</span>, i.e. the number of variants where both samples have non-missing genotypes. While this is arguably a better estimator of the true GRM (trading shrinkage for noise), it has the drawback that one loses the clean interpretation of the loadings and scores as features and projections.</p>
<p>Separately, for the PCs PLINK/GCTA output the eigenvectors of the GRM; even ignoring the above discrepancy that means the left singular vectors <span class="math inline">\(U_k\)</span> instead of the component scores <span class="math inline">\(U_k S_k\)</span>. While this is just a matter of the scale on each PC, the scores have the advantage of representing true projections of the data onto features with the variance of a score reflecting the variance explained by the corresponding feature. (In PC bi-plots this amounts to a change in aspect ratio; for use of PCs as covariates in regression it is immaterial.)</p>
</body>
</html>
